### PR TITLE
Workaround for WSL1 process IO redirection bug

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Attempts to implement a workaround for #40727. Alternative to #40832.

Has been tested with WSL1 and WSL2.

### CODE REVIEW

**Comment by danmosemsft** [comment]
Has anyone run all our unit tests on WSL1 recently? Not sure whether this issue is protected by a test but we didn't run on WSL1, or not, but if we haven't run them all it would be good to do so...

**Comment by wfurt** [comment]
not that I know of. @ManickaP did some investigations while back for networking and we decided not to fix the issues and recommend WSL2.

**Comment by eiriktsarpalis** [comment]
@danmosemsft It's currently impossible to build dotnet/runtime on WSL1, due to this bug impairing the bootstrapping SDK. After manually patching the bootstrapping sdk, I'm seeing failing tests that are unrelated to this issue, though I have only focused on the impacted projects (sockets & process) and haven't run the entire suite yet.

**Review by jeffhandley — COMMENTED** [review]
Very cool, @eiriktsarpalis. I cannot sign off, but with a domain expert approval, I'd support taking the fix instead of reverting. Thanks for the effort here.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs [hunk 1]
```diff
@@ -761,7 +761,23 @@ internal static TimeSpan TicksToTimeSpan(double ticks)
         private static Stream OpenStream(int fd, FileAccess access)
         {
             Debug.Assert(fd >= 0);
-            var socket = new Socket(new SafeSocketHandle((IntPtr)fd, ownsHandle: true));
+            var socketHandle = new SafeSocketHandle((IntPtr)fd, ownsHandle: true);
+            var socket = new Socket(socketHandle);
+
+            if (!socket.Connected)
+            {
+                // WSL1 workaround -- due to issues with sockets syscalls
```
Comment (am11) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:769]:
```suggestion
                // WSL1 workaround -- due to issues with sockets syscalls
```
will fix the build failure 🤣

**Review by wfurt — APPROVED** [review]
Looks reasonable. I assume the new alternative is essentially what it was befe, right?

**Comment by danmosemsft** [comment]
> we decided not to fix the issues and recommend WSL2

For development, sure, but it seems Powershell care about WSL1 ..

**Comment by stephentoub** [comment]
The workaround LGTM. But the Socket(SafeSocketHandle) ctor still ends up being broken in WSL; any other use is going to run into similar problems.  It'd still be good to dive deeper into the ctor to understand exactly what's failing where.

**Comment by wfurt** [comment]
We should open separate issue for that and we can investigate after 5. Since that is new API there should not be any existing users.

**Comment by eiriktsarpalis** [comment]
> I assume the new alternative is essentially what it was befe, right?

The only difference is that it's using socket pairs instead of anonymous pipes: https://github.com/dotnet/runtime/pull/34861/files#diff-df5b2051e73600a29fbe0b3230daf7d1L276-R302

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs [hunk 2]
```diff
@@ -761,7 +761,23 @@ internal static TimeSpan TicksToTimeSpan(double ticks)
...
+                // socket pairs fd's are erroneously inferred as not connected.
+                // Fall back to using FileStream instead.
+
+                GC.SuppressFinalize(socket);
+                GC.SuppressFinalize(socketHandle);
+
+                return new FileStream(
```
Comment (eiriktsarpalis) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:776]: A thing to note here is that the particular fallback will result in the `ReadAsync_OutputStreams_Cancel_RespondsQuickly` test to fail.

**Comment by davidfowl** [comment]
We use the new socket API in kestrel.

**Comment by eiriktsarpalis** [comment]
There are a few builds timing out, which seems to happen in other PRs as well.

**Pull request merged**